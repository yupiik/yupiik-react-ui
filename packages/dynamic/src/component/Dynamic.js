import { createContext, h } from 'react';
import { HtmlRegistry } from './registry';

const getErrorComponent = (registry) => registry.error || (({ children }) => (<div className="error">{children})</div>));

const NoComponentFound = (registry, name) => {
    const Error = getErrorComponent(registry);
    return (
        <Error name={name || ''}>No component {name || '?'} found!</Error>
    );
}

/**
 * A generic component using a component registry to render the request (config).
 * @param {object} registry the component registry (an object where the key is component name).
 * @param {object} config the underlying component configuration (name and options).
 * @returns a preact component which looks up in the registry the actual component to use for rendering.
 */
export const Dynamic = ({
    registry,
    options: { name, options = {} },
}) => {
    const contextualRegistry = useContext(ComponentRegistryContext);
    const internalRegistry = registry || contextualRegistry || {};

    if (!name) {
        return (<NoComponentFound registry={internalRegistry} name={name} />);
    }

    let Component = internalRegistry[name];
    if (!Component) { // fallback by default
        Component = HtmlRegistry[name];
    }
    if (!Component) {
        return (<NoComponentFound registry={internalRegistry} name={name} />);
    }

    const { children, ...rest } = options || {};
    return (
        <Component
            registry={internalRegistry}
            {...(rest || {})}>
            {children}
        </Component >
    );
};

export const ComponentRegistryContext = createContext();
