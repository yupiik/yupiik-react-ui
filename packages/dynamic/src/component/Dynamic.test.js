import { render } from '@testing-library/react';
import { expect } from 'expect';
import { ComponentRegistryContext, Dynamic } from './Dynamic';


const Container = ({ children, registry }) => (
    <div className="container">
        {children
            .map(options => (
                <Dynamic
                    registry={registry}
                    options={options}
                />))}
    </div>
);
const Greeting = ({ name }) => (<h1>Hello {name}!</h1>);
const Text = ({ value }) => (<div>{value}</div>);

const registry = {
    container: Container,
    greeting: Greeting,
    text: Text,
};

test('Dynamic should render', () => {
    const { container: { children } } = render(
        <Dynamic
            registry={registry}
            options={{
                name: 'greeting',
                options: {
                    name: 'Tests',
                },
            }}
        />
    );
    expect(children).toMatchSnapshot();
});

test('Dynamic should render with contextual registry', () => {
    expect(render(
        <ComponentRegistryContext.Provider value={registry}>
            <Dynamic
                options={{
                    name: 'greeting',
                    options: {
                        name: 'Tests',
                    },
                }}
            />
        </ComponentRegistryContext.Provider>
    ).container.children).toMatchSnapshot();
});

test('Dynamic should render complex trees', () => {
    const { container: { children } } = render(
        <Dynamic
            registry={registry}
            options={{
                name: 'container',
                options: {
                    children: [
                        {
                            name: 'greeting',
                            options: {
                                name: 'Yupiik',
                            },
                        },
                        {
                            name: 'text',
                            options: {
                                value: 'Some content.'
                            },
                        },
                    ],
                },
            }}
        />
    );
    expect(children).toMatchSnapshot();
});

test('Dynamic should render component not found', () => {
    expect(
        render(
            <Dynamic
                registry={registry}
                options={{ name: 'missing' }}
            />
        ).container.children
    ).toMatchSnapshot();
});

test('Dynamic no registry', () => {
    expect(() => render(
        <Dynamic
            options={{ name: 'greeting' }}
        />
    )).toThrow();
});
