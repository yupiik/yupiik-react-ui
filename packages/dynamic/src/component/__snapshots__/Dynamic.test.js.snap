// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dynamic should render 1`] = `
HTMLCollection [
  <h1>
    Hello 
    Tests
    !
  </h1>,
]
`;

exports[`Dynamic should render complex trees 1`] = `
HTMLCollection [
  <div
    class="container"
  >
    <h1>
      Hello 
      Yupiik
      !
    </h1>
    <div>
      Some content.
    </div>
  </div>,
]
`;

exports[`Dynamic should render component not found 1`] = `
HTMLCollection [
  <div
    class="error"
  >
    No component 
     found!
    )
  </div>,
]
`;

exports[`Dynamic should render with contextual registry 1`] = `
HTMLCollection [
  <h1>
    Hello 
    Tests
    !
  </h1>,
]
`;
